ads/google/a4a/utils.js

```diff
@@ -293,3 +280,62 @@ export function getCorrelator(win, opt_cid) {
   }
   return win.ampAdPageCorrelator;
 }
+
+/*
+ * Browser viewport size, if we are in an iframe.
+ * @param {!Window} win The window for which we read the browser dimensions.
+ * @return {?{width: number, height: number}}
+ */
+function browserViewportSize(win) {
+  const w = win.top;
+  if (win != w) {
+    let width = -1;
+    let height = -1;
+    try {
+      if (w.document && w.document.body) {
+        const body = w.document.body;
+        width = Math.round(body./*OK*/clientWidth);
+        height = Math.round(body./*OK*/clientHeight);
+      }
+    } catch (e) {
+      width = -0xbadbad;
+      height = -0xbadbad;
+    }
+    return {width, height};
+  };
+  return null;
+}
+
+/**
+ * Collect additional dimensions for the brdim parameter.
+ * @param {!Window} win The window for which we read the browser dimensions.
+ * @param {!../../../src/service/viewport-impl.Viewport} viewport
+ * @return {string}
+ */
+function additionalDimensions(win, viewport) {
+  // Some browsers throw errors on some of these.
+  let screenX, screenY, outerWidth, outerHeight, innerWidth, innerHeight;
+  const size = viewport.getSize();
+  try {
+    screenX = win.screenX;
+    screenY = win.screenY;
+  } catch (e) {}
+  try {
+    outerWidth = win.outerWidth;
+    outerHeight = win.outerHeight;
+  } catch (e) {}
+  try {
+    innerWidth = size.width;
+    innerHeight = size.height;
+  } catch (e) {}
+  return [win.screenLeft,
+          win.screenTop,
+          screenX,
+          screenY,
+          win.screen ? win.screen.availWidth : undefined,
+          win.screen ? win.screen.availTop : undefined,
+          outerWidth,
+          outerHeight,
+          innerWidth,
+          innerHeight].join();
+};
```

Comment:
I added a test for this function, and a TODO for others that I am not touching in this PR.